### DISCUSSION TITLE
how to trap keypress inside a command?

### DISCUSSION THREAD

**Discussion by elbrandt** [daniele77/cli#222] [Q&A]
Hi,
Fantastic library, well written and feature packed. Thanks for your work.
We're using CliLocalTerminalSession with a LoopScheduler. One of the commands we've implemented, we'd like the command to run in a loop 'until the user presses a key' to end the command and return to the menu.

When we implement this in our command's lambda using the `select()` function, they way we typically would in a regular command line app, I find that we're in a race condition with a thread running the same `select()` command in an instance of the LinuxKeyboard class. So, I think either we're doing it wrong, or this "run command until key is pressed" use case isn't supported in this library.
Is there a relatively pain-free way to do what we want? Can we somehow temporarily pause the scheduler within our command?
I can elaborate more on the use case if this question is unclear.
Thanks again for an excellent library.
-Eric.

**Comment by daniele77** [comment]
Hi @elbrandt 
first of all, thank you very much for using our library.
If you like it, please consider donating using the heart icon on top of this page :-)

Supposing you're on a Linux machine, you can try something like this:


    #include <thread>
    #include <atomic>
    #include <termios.h>
    #include <unistd.h>

    ...
        
    menu->Insert(
            "cycle",
            [](std::ostream& out)
            { 
                out << "Press any key to stop the cycle\n";
                 std::atomic<bool> stop(false);
                 std::thread servant([&]()
                {
                    while (!stop)
                    {
                        out << "Cycling...\n";
                        std::this_thread::sleep_for(std::chrono::seconds(1));
                    }
                });

                struct termios t;
                constexpr tcflag_t ICANON_FLAG = ICANON;
                constexpr tcflag_t ECHO_FLAG = ECHO;                    
                // get current setup
                tcgetattr(STDIN_FILENO, &t);
                // disable canonic input and echo
                t.c_lflag &= ~(ICANON_FLAG | ECHO_FLAG);
                // update new settings...
                tcsetattr(STDIN_FILENO, TCSANOW, &t);                    

                getchar();

                 stop = true;
                 servant.join();
            },
            "Start a cycle in another thread until any key is pressed"
    );

I've tested the code and everything works.
Of course, this solution cannot work on remote consoles, but since you're using `CliLocalTerminalSession `, this should not represent a problem for your use case.

**Reply by elbrandt** [reply]
Thanks! We'll try this out ASAP. I appreciate the versatility of your library to accommodate this.

**Comment by daniele77** [comment]
I've just committed a fix that should enable user keyboard handling on Linux platform as well (commit 17dafd8).
It's the issue #239.
Now you can use `std::cin` and `std::getline` inside command handlers (local session only, of course).
Please, see if this solves your problem.
